# BulkRNA References
- **adapters.fa** - fasta file containing adapter sequences, used for trimming reads
- **aducanumab.yaml** - yaml file to recreate conda environment used for fastq preprocessing
- **bulkRNA_SRA_metadata.tsv** - SRA run accessions with sample metadata
- **config.json** - Snakemake config file, created by running python script in scRNA/scripts/create_config.py
- **core_metadata.tsv** - metadata from Mayo Clinic Genome Anlaysis Core
- **fastq_file_list.txt** - list of all fastq files
- **fastq_file_headers.tsv** - headers of all fastq files, used to check for batch effect
- **metadata.tsv** - the metadata file used for the R analysis, this file is the result of merging core_metadata.tsv and mika_metadata.tsv, it does not contain the SRA run accessions like in bulkRNA_SRA_metadata.tsv
- **mika_metadata.tsv** - metadata provided by Mika on the experiment setup
- **salmon.yml** - conda environment only containing the tool salmon, used to validate the library type of the fastq files
- **sample_file_list.txt** - list of all samples (there are multiple fastq files per sample)
